extensions/script/lua/LuaBaseStream.cpp

```diff
@@ -52,16 +52,11 @@ std::string LuaBaseStream::read(size_t len) {
   //     0 <= n < s.size()."
   //
   // http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf
-  auto read = stream_->read(reinterpret_cast<uint8_t *>(&buffer[0]), static_cast<int>(len));
-  if (read < 0) {
-    return nullptr;
-  }
-
-  if (gsl::narrow<size_t>(read) != len) {
-    buffer.resize(gsl::narrow<size_t>(read));
+  const auto read = stream_->read(reinterpret_cast<uint8_t *>(&buffer[0]), len);
+  if (!io::isError(read) && read != len) {
+    buffer.resize(read);
   }
-
-  return buffer;
+  return io::isError(read) ? std::string{} : buffer;
```

Comment:
We used to throw an exception (in an obfuscated way) in the error case, now we return an empty string.  This is probably good, but I don't know if we want to make a change like this in a refactoring PR.  If we do want to make this change, we should probably change line 42, too.

Reply:
My understanding is that the old version returned `nullptr` as `std::string`, invoking the `std::string::string(const char*)` constructor which led to undefined behavior in case of an error. I thought replacing this with a defined empty string would be better, but I might be missing something.

Reply:
This snippet was changed during the rebase, btw. Previously it was a simple comparison + resize without the `narrow` and the `if (read < 0)` condition

Reply:
yup, I said "probably good", you said "would be better", there is no disagreement there :)

libstdc++ throws an exception, but you're right that other implementations may behave differently

my point is that the same thing happens at line 42

Reply:
I didn't know that libstdc++ throws there. Despite being an unrelated change, I really don't want to rely on undefined behavior there, so I insist on a change.

In the old main version (that this PR was originally based on), there was no string construction from nullptr here (but the one on line 42 was already present). I believe that change on main to be an error, so by moving to returning an empty string, we basically revert to the old, correct version, the only change being that we no longer return an empty string of size `len`.

If you insist on throwing, I can change this to throw something explicitly, but I prefer the old behavior of returning an empty string. I want to change the occurrence on line 42 to be consistent with this one, too.

Reply:
Yes, I think returning an empty string at line 42 would be fine.